keylime/push_agent_monitor.py

```diff
@@ -0,0 +1,244 @@
+"""Monitor PUSH mode agents for timeout detection.
+
+This module implements event-driven timeout detection for PUSH mode agents.
+Instead of polling the database continuously, it schedules individual timeout
+callbacks when attestations are received. When an agent exceeds the timeout
+threshold, its accept_attestations flag is set to False, which causes
+attestation_status to show as "FAIL".
+
+Integration points:
+- Call schedule_agent_timeout(agent_id) when a PUSH mode attestation is received
+- Call cancel_agent_timeout(agent_id) when a PUSH mode agent is deleted
+
+This approach eliminates continuous database polling, significantly reducing load.
+"""
+
+import time
+from typing import Dict, Optional
+
+import tornado.ioloop
+from sqlalchemy import and_, or_
+
+from keylime import config, keylime_logging
+from keylime.db.keylime_db import SessionManager, make_engine
+from keylime.db.verifier_db import VerfierMain
+
+logger = keylime_logging.init_logging("push_agent_monitor")
+
+# Timeout multiplier for quote_interval
+# We wait 2x the quote_interval before marking an agent as failed
+# This allows for network delays and processing time
+_TIMEOUT_MULTIPLIER = 2.0
+
+# In-memory map of agent_id -> timeout handle for active timeouts
+# This allows us to cancel/reschedule timeouts without database queries
+_agent_timeout_handles: Dict[str, object] = {}
```

Comment:
We access this global from many places, e.g. in `_mark_agent_failed()` (`_agent_timeout_handles.pop()`), in `schedule_agent_timeout()` (`_agent_timeout_handles[agent_id] = timeout_handle`) and `cancel_agent_timeout()` (`timeout_handle = _agent_timeout_handles.pop(agent_id, None)`) and there is no syncrhonization. Since tornado requests run in different contexts than IOLoop callbacks, this could be an issue. Also, multiple attestation could arrive simultaneously for different agents.

Please add a lock to protect all access to this dictionary.